File: archived/Geolocation/js/geolocationjs/js/scenario3_BackgroundTask.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved
//// This code is licensed under the MIT License (MIT).

(function () {
    "use strict";

    var Background = Windows.ApplicationModel.Background;
    var geolocTask;
    var sampleBackgroundTaskName = "SampleLocationBackgroundTask";
    var sampleBackgroundTaskEntryPoint = "js\\backgroundtask.js";
    var disposed;

    var page = WinJS.UI.Pages.define("/html/scenario3_BackgroundTask.html", {
        ready: function (element, options) {
            disposed = false;
            document.getElementById("registerBackgroundTaskButton").addEventListener("click", registerBackgroundTask, false);
            document.getElementById("unregisterBackgroundTaskButton").addEventListener("click", unregisterBackgroundTask, false);

            // Loop through all background tasks to see if our task is already registered
            var iter = Windows.ApplicationModel.Background.BackgroundTaskRegistration.allTasks.first();
            var hascur = iter.hasCurrent;
            while (hascur) {
                var cur = iter.current.value;
                if (cur.name === sampleBackgroundTaskName) {
                    geolocTask = cur;
                    break;
                }
                hascur = iter.moveNext();
            }

            if (geolocTask) {

                // Associate an event handler to the existing background task
                geolocTask.addEventListener("completed", onCompleted);
                try {
                    var backgroundAccessStatus = Background.BackgroundExecutionManager.getAccessStatus();
                    switch (backgroundAccessStatus) {
                        case Background.BackgroundAccessStatus.alwaysAllowed:
                        case Background.BackgroundAccessStatus.allowedSubjectToSystemPolicy:
                            WinJS.log && WinJS.log("Background task is already registered. Waiting for next update...", "sample", "status");
                            break;

                        default:
                            WinJS.log && WinJS.log("Background tasks may be disabled for this app", "sample", "status");
                            break;
                    }
                } catch (ex) {
                    WinJS.log && WinJS.log(ex.toString(), "sample", "error");
                }

                TimerTriggerBackgroundTask.updateButtonStates(/*registered:*/ true);
            } else {
                TimerTriggerBackgroundTask.updateButtonStates(/*registered:*/ false);
            }
        },
        dispose: function () {
            if (!disposed) {
                disposed = true;

                if (geolocTask) {
                    geolocTask.removeEventListener("completed", onCompleted);
                }
            }
        }
    });

    // Handle background task completion
    function onCompleted() {
        try {
            var settings = Windows.Storage.ApplicationData.current.localSettings;

            WinJS.log && WinJS.log(settings.values["Status"], "sample", "status");

            document.getElementById("latitude").innerText = settings.values["Latitude"];
            document.getElementById("longitude").innerText = settings.values["Longitude"];
            document.getElementById("accuracy").innerText = settings.values["Accuracy"];
        } catch (ex) {
            WinJS.log && WinJS.log(ex.toString(), "sample", "error");
        }
    }

    function registerBackgroundTask() {
        try {
            // Get permission for a background task from the user. If the user has already answered once,
            // this does nothing and the user must manually update their preference via PC Settings.
            Background.BackgroundExecutionManager.requestAccessAsync().done(
                function (backgroundAccessStatus) {
                    var builder = new Windows.ApplicationModel.Background.BackgroundTaskBuilder();

                    // Regardless of the answer, register the background task. If the user later adds this application
                    // to the lock screen, the background task will be ready to run.

                    // Create a new background task builder
                    builder.name = sampleBackgroundTaskName;
                    builder.taskEntryPoint = sampleBackgroundTaskEntryPoint;

                    // Create a new timer triggering at a 15 minute interval
                    // Associate the timer trigger with the background task builder
                    builder.setTrigger(new Windows.ApplicationModel.Background.TimeTrigger(15, false));

                    // Register the background task
                    geolocTask = builder.register();

                    // Associate an event handler with the new background task
                    geolocTask.addEventListener("completed", onCompleted);

                    TimerTriggerBackgroundTask.updateButtonStates(/*registered:*/ true);

                    switch (backgroundAccessStatus) {
                        case Background.BackgroundAccessStatus.alwaysAllowed:
                        case Background.BackgroundAccessStatus.allowedSubjectToSystemPolicy:
                            WinJS.log && WinJS.log("Background task registered.", "sample", "status");

                            // Need to request access to location
                            // This must be done with the background task registration
                            // because the background task cannot display UI.
                            requestLocationAccess();
                            break;

                        default:
                            WinJS.log && WinJS.log("Background tasks may be disabled for this app", "sample", "error");
                            break;
                    }
                },
                function (e) {
                    // Did you forget to do the background task declaration in the package manifest?
                    WinJS.log && WinJS.log(e.toString(), "sample", "error");
                }
            );
        } catch (ex) {
            WinJS.log && WinJS.log(ex.toString(), "sample", "error");
        }
    }

    function unregisterBackgroundTask() {
        // Unregister the background task
        if (geolocTask) {
            geolocTask.unregister(true);
            geolocTask = null;
        }

        WinJS.log && WinJS.log("Background task unregistered", "sample", "status");

        document.getElementById("latitude").innerHTML = "No data";
        document.getElementById("longitude").innerHTML = "No data";
        document.getElementById("accuracy").innerHTML = "No data";

        TimerTriggerBackgroundTask.updateButtonStates(/*registered:*/ false);
    }

    function requestLocationAccess() {
        Windows.Devices.Geolocation.Geolocator.requestAccessAsync().done(
        function (accessStatus) {
            switch (accessStatus) {
                case Windows.Devices.Geolocation.GeolocationAccessStatus.allowed:
                    break;

                case Windows.Devices.Geolocation.GeolocationAccessStatus.denied:
                    WinJS.log && WinJS.log("Access to location is denied.", "sample", "error");
                    break;

                case Windows.Devices.Geolocation.GeolocationAccessStatus.unspecified:
                    WinJS.log && WinJS.log("Unspecified error!", "sample", "error");
                    break;
            }
        },
        function (err) {
            WinJS.log && WinJS.log(err, "sample", "error");
        });
    }
})();

// Update the disabled state of the register/unregister buttons.
var TimerTriggerBackgroundTask = {
    "updateButtonStates": function (registered) {
        var registerButton = document.getElementById("registerBackgroundTaskButton");
        var unregisterButton = document.getElementById("unregisterBackgroundTaskButton");

        registerButton && (registerButton.disabled = registered);
        unregisterButton && (unregisterButton.disabled = !registered);
    }
};

File: Samples/Geolocation/cppwinrt/Scenario3_BackgroundTask.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario3_BackgroundTask.h"
#include "Scenario3_BackgroundTask.g.cpp"
#include "SampleConfiguration.h"

using namespace winrt;
using namespace winrt::Windows::ApplicationModel::Background;
using namespace winrt::Windows::Devices::Geolocation;
using namespace winrt::Windows::Foundation;
using namespace winrt::Windows::Storage;
using namespace winrt::Windows::UI::Xaml;
using namespace winrt::Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario3_BackgroundTask::Scenario3_BackgroundTask()
    {
        InitializeComponent();
    }

    void Scenario3_BackgroundTask::OnNavigatedTo(NavigationEventArgs const&)
    {
       // See if our background task is already registered.
        m_geolocTask = LookupBackgroundTask(BackgroundTaskName);

        if (m_geolocTask != nullptr)
        {
            // Associate an event handler with the existing background task
            m_taskCompletedToken = m_geolocTask.Completed({ get_weak(), &Scenario3_BackgroundTask::OnCompleted });
        }
        UpdateButtonStates();
    }

    void Scenario3_BackgroundTask::OnNavigatedFrom(NavigationEventArgs const&)
    {
        if (m_geolocTask != nullptr)
        {
            // Remove the event handler
            m_geolocTask.Completed(m_taskCompletedToken);
        }
    }

    void Scenario3_BackgroundTask::RegisterBackgroundTask(IInspectable const&, RoutedEventArgs const&)
    {
        // Register the background task without checking whether the user has enabled background execution.
        // If it's disabled, and then the user later enables background execution, the background task will be ready to run.
        // Create a new background task builder
        BackgroundTaskBuilder geolocTaskBuilder;

        geolocTaskBuilder.Name(BackgroundTaskName);
        geolocTaskBuilder.TaskEntryPoint(BackgroundTaskEntryPoint);

        // Create a new timer triggering at a 15 minute interval
        TimeTrigger trigger(15, false);

        // Associate the timer trigger with the background task builder
        geolocTaskBuilder.SetTrigger(trigger);

        // Register the background task
        m_geolocTask = geolocTaskBuilder.Register();

        // Associate an event handler with the new background task
        m_taskCompletedToken = m_geolocTask.Completed({ get_weak(), &Scenario3_BackgroundTask::OnCompleted });

        UpdateButtonStates();
        CheckBackgroundAndRequestLocationAccess();
    }

    void Scenario3_BackgroundTask::UnregisterBackgroundTask(IInspectable const&, RoutedEventArgs const&)
    {
        // Unregister the background task
        if (m_geolocTask != nullptr)
        {
            m_geolocTask.Unregister(true);
            m_geolocTask = nullptr;
        }

        ScenarioOutput_Latitude().Text(L"No data");
        ScenarioOutput_Longitude().Text(L"No data");
        ScenarioOutput_Accuracy().Text(L"No data");
        UpdateButtonStates();
        m_rootPage.NotifyUser(L"Background task unregistered", NotifyType::StatusMessage);
    }

    // This event handler is called when the background task is completed.
    fire_and_forget Scenario3_BackgroundTask::OnCompleted(IBackgroundTaskRegistration sender, BackgroundTaskCompletedEventArgs e)
    {
        auto lifetime = get_strong();

        // Update the UI with progress reported by the background task
        co_await resume_foreground(Dispatcher());

        try
        {
            // Rethrow any exception that occurred in the background task.
            e.CheckResult();

            // Update the UI with the completion status of the background task
            // The Run method of the background task sets this status.
            ReportSavedStatus();

            // Extract and display location data set by the background task if not null
            auto values = ApplicationData::Current().LocalSettings().Values();
            ScenarioOutput_Latitude().Text(unbox_value_or(values.TryLookup(L"Latitude"), L"No data"));
            ScenarioOutput_Longitude().Text(unbox_value_or(values.TryLookup(L"Longitude"), L"No data"));
            ScenarioOutput_Accuracy().Text(unbox_value_or(values.TryLookup(L"Accuracy"), L"No data"));
        }
        catch (...)
        {
            // The background task had an error
            m_rootPage.NotifyUser(to_message(), NotifyType::ErrorMessage);
        }
    }

    // Update the enable state of the register/unregister buttons
    // based on whether we have already registered a task.
    void Scenario3_BackgroundTask::UpdateButtonStates()
    {
        bool registered = (m_geolocTask != nullptr);
        RegisterBackgroundTaskButton().IsEnabled(!registered);
        UnregisterBackgroundTaskButton().IsEnabled(registered);
    }
}
